ccm_web/server/src/api/api.enroll.section.users.handler.ts

```diff
@@ -0,0 +1,99 @@
+import CanvasRequestor from '@kth/canvas-api'
+
+import { APIErrorData, isAPIErrorData } from './api.interfaces'
+import { handleAPIError, HttpMethod } from './api.utils'
+
+import baseLogger from '../logger'
+import { CanvasEnrollment } from '../canvas/canvas.interfaces'
+import { SectionUserDto } from './dtos/api.section.users.dto'
+
+const logger = baseLogger.child({ filePath: __filename })
+
+export class EnrollSectionUsersApiHandler {
+  requestor: CanvasRequestor
+  users: SectionUserDto[]
+  sectionId: number
+
+  constructor (requestor: CanvasRequestor, users: SectionUserDto[], sectionId: number) {
+    this.requestor = requestor
+    this.users = users
+    this.sectionId = sectionId
+  }
+
+  makeResponse (enrollmentResult: Array<APIErrorData | CanvasEnrollment>): CanvasEnrollment[] | APIErrorData {
+    const failures = []
+    const statusCodes: Set<number> = new Set()
+    const successes = []
+
+    for (const enrollment of enrollmentResult) {
+      if (isAPIErrorData(enrollment)) {
+        const {
+          statusCode,
+          errors
+        } = enrollment
+        failures.push(...errors)
+        statusCodes.add(statusCode)
+      } else {
+        successes.push(enrollment)
+      }
+    }
+
+    if (successes.length === this.users.length) {
+      return successes
+    } else {
+      return {
+        statusCode: statusCodes.size > 1 ? 502 : [...statusCodes][0],
+        errors: failures
+      }
+    }
+  }
+
+  async enrollUser (user: SectionUserDto): Promise<CanvasEnrollment | APIErrorData> {
+    try {
+      const endpoint = `sections/${this.sectionId}/enrollments`
+      const method = HttpMethod.Post
+      const body = {
+        enrollment: {
```

Comment:
... ahem ...

Reply:
I'm sorry, this was not the nicest way to draw attention to this. Just another data point in the `user` versus `enrollment` discussion I suppose.

Reply:
My apologies.  I was not intentionally ignoring the discussion about "user" vs. "enrollment".  I'm just intensely focused on handling the Canvas API's "rate limit exceeded" error.  That seems to be my most important task at the moment.

Perhaps I misspoke about the input values required by Canvas API at the time I wrote my comment.  Or maybe I think about it from a different perspective than you do.  I can rename the symbols to whatever the group consensus is.  I guess the naming convention doesn't really matter to me in comparison to the error handling.

Looking at the section of code you cited, the lines that follow it are significant, too…

```js
      const body = {
        enrollment: {
          user_id: `sis_login_id:${user.loginId}`,
          type: user.type,
```

We're asking Canvas to make an enrollment based on the ID and type of a user.  We give Canvas info about a user (from a CSV file of users) and it makes an enrollment.

I mention this to respond to your point and to show what my thinking was when I named the symbols.  If I'm using inconsistent or confusing naming, I'm willing to change it.

Although not entirely applicable, our conversation reminds me of this quote:

> "Substitute 'damn' every time you're inclined to write 'very;' your editor will delete it and the writing will be just as it should be." ― Mark Twain

😉

Reply:
Yeah, again, I'm leaving this up to your discretion. Variable/method naming is important, in my opinion, but less important than error handling (as you say) since the names are OK at the moment.

All that said, if you think about it, the `type` value you provide has nothing inherently to do with the `user` entity; it's describing the relation between the `user` and the `course`/`section`, i.e. the `enrollment`. The `type` value is not scoped to all of Canvas, it's scoped to a particular context, changing from section to section. So you're really providing more than a `user` to the API.

Reply:
Yes, I can imagine it that way, too.  It's an enrollment type, because each valid type includes that word in it, like `StudentEnrollment`, etc.

So, I'll keep an open mind about this.  After the rate limit problem is under control, I promise to revisit this.  Maybe by then, I'll look at this with fresher eyes.

---

Mel: What do you plan to do today?
Carl: First, I'm going to buy some new glasses.
Mel: What will you do after that?
Carl: Then, I'll see.

Reply:
So, I've thought about this and I can see your point of view, but I also think the way I've named it is valid, too.

So, I'm going to leave it as is.  If someone wants to change it in the future… 🤷‍♂️